File: packages/protocol/src/stateMachine.ts
import { MessageStatus, isExpired, type HopMessage } from "./message.js";

export class IllegalStateTransitionError extends Error {
  constructor(
    public readonly from: MessageStatus,
    public readonly to: MessageStatus,
  ) {
    super(`Illegal message transition: ${from} -> ${to}`);
    this.name = "IllegalStateTransitionError";
  }
}

export const ALLOWED_TRANSITIONS: Record<MessageStatus, readonly MessageStatus[]> = {
  CREATED: [
    MessageStatus.ENCRYPTING,
    MessageStatus.ENCRYPTED,
    MessageStatus.FAILED,
    MessageStatus.EXPIRED,
  ],
  ENCRYPTING: [MessageStatus.ENCRYPTED, MessageStatus.FAILED, MessageStatus.EXPIRED],
  ENCRYPTED: [MessageStatus.QUEUED, MessageStatus.FAILED, MessageStatus.EXPIRED],
  QUEUED: [
    MessageStatus.SENDING,
    MessageStatus.RETRYING,
    MessageStatus.FAILED,
    MessageStatus.EXPIRED,
  ],
  RETRYING: [MessageStatus.SENDING, MessageStatus.FAILED, MessageStatus.EXPIRED],
  SENDING: [
    MessageStatus.SENT,
    MessageStatus.QUEUED,
    MessageStatus.RETRYING,
    MessageStatus.FAILED,
    MessageStatus.EXPIRED,
  ],
  SENT: [
    MessageStatus.DELIVERED,
    MessageStatus.RELAYING,
    MessageStatus.FAILED,
    MessageStatus.EXPIRED,
  ],
  RELAYING: [
    MessageStatus.DELIVERED,
    MessageStatus.RELAYING,
    MessageStatus.FAILED,
    MessageStatus.EXPIRED,
  ],
  DELIVERED: [MessageStatus.READ],
  READ: [],
  FAILED: [MessageStatus.QUEUED],
  EXPIRED: [],
};

export function canTransition(from: MessageStatus, to: MessageStatus): boolean {
  return ALLOWED_TRANSITIONS[from].includes(to);
}

export function transition(message: HopMessage, to: MessageStatus, now = new Date()): HopMessage {
  if (isExpired(message, now) && to !== MessageStatus.EXPIRED) {
    throw new IllegalStateTransitionError(message.status, to);
  }
  if (!canTransition(message.status, to)) {
    throw new IllegalStateTransitionError(message.status, to);
  }
  return { ...message, status: to };
}

File: apps/api/app/messaging/state_machine.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

MessageStatus = Literal[
    "CREATED",
    "ENCRYPTING",
    "ENCRYPTED",
    "QUEUED",
    "RETRYING",
    "SENDING",
    "SENT",
    "RELAYING",
    "DELIVERED",
    "READ",
    "FAILED",
    "EXPIRED",
]

ALLOWED_TRANSITIONS: dict[MessageStatus, tuple[MessageStatus, ...]] = {
    "CREATED": ("ENCRYPTING", "ENCRYPTED", "FAILED", "EXPIRED"),
    "ENCRYPTING": ("ENCRYPTED", "FAILED", "EXPIRED"),
    "ENCRYPTED": ("QUEUED", "FAILED", "EXPIRED"),
    "QUEUED": ("SENDING", "RETRYING", "FAILED", "EXPIRED"),
    "RETRYING": ("SENDING", "FAILED", "EXPIRED"),
    "SENDING": ("SENT", "QUEUED", "RETRYING", "FAILED", "EXPIRED"),
    "SENT": ("DELIVERED", "RELAYING", "FAILED", "EXPIRED"),
    "RELAYING": ("DELIVERED", "RELAYING", "FAILED", "EXPIRED"),
    "DELIVERED": ("READ",),
    "READ": (),
    "FAILED": ("QUEUED",),
    "EXPIRED": (),
}


class IllegalStateTransitionError(ValueError):
    def __init__(self, src: MessageStatus, dest: MessageStatus) -> None:
        super().__init__(f"Illegal message transition: {src} -> {dest}")
        self.src = src
        self.dest = dest


def can_transition(src: MessageStatus, dest: MessageStatus) -> bool:
    return dest in ALLOWED_TRANSITIONS[src]


def transition(src: MessageStatus, dest: MessageStatus) -> MessageStatus:
    if not can_transition(src, dest):
        raise IllegalStateTransitionError(src, dest)
    return dest


def should_stop_forwarding(hop_count: int, expires_at: datetime, now: datetime, max_hops: int = 8) -> bool:
    return hop_count >= max_hops or now >= expires_at
